<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • genomicMateSelectR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genomicMateSelectR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/non_additive_models.html">Genomic predictions with non-additive effects</a>
    </li>
    <li>
      <a href="../articles/start_here.html">Getting started predicting crosses</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/wolfemd/genomicMateSelectR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/wolfemd/genomicMateSelectR/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.0" id="genomicmateselectr-020">genomicMateSelectR 0.2.0<a class="anchor" aria-label="anchor" href="#genomicmateselectr-020"></a></h2>
<p>Upgrade. Combine <strong>predCrossVar</strong> with other functions for GS pipeline <em>and</em> add high level wrapping functions that implement/automatic cross predictions, and more.</p>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.0" id="genomicmateselectr-010">genomicMateSelectR 0.1.0<a class="anchor" aria-label="anchor" href="#genomicmateselectr-010"></a></h2>
<p>See <a href="https://wolfemd.github.io/predCrossVar/" class="external-link">predCrossVar</a>, the OG codebase.</p>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.0" id="genomicmateselectr-000">genomicMateSelectR 0.0.0<a class="anchor" aria-label="anchor" href="#genomicmateselectr-000"></a></h2>
<div class="section level3">
<h3 id="development-to-do-0-0-0">Development To Do<a class="anchor" aria-label="anchor" href="#development-to-do-0-0-0"></a></h3>
<ul class="task-list"><li><p><input type="checkbox" disabled>
Improve input format for SNP-effects into <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>. More flexible SNP-effect list-column naming (instead of opinionated e.g. <strong>allelesubeffectlist</strong>)? Particularly for the <strong>DirDom</strong> model.</p></li>
<li>
<p><input type="checkbox" disabled>
Improve phenotype input e.g. for <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> and <code><a href="../reference/runParentWiseCrossVal.html">runParentWiseCrossVal()</a></code>.</p>
<ul><li>Currently demands a number of variables, be named in particular weights, owing to the two-stage genomic prediction process I use for NextGen Cassava. In that Cassava pipeline, I’ve used de-regressed BLUPs as response variable for genomic prediction and weighted error variances according to a weighting scheme devised by <em>Garrick et al. 2009</em>. Users can subvert this, but are currently required to use certain naming: <strong>drgBLUP</strong> for the response data, <strong>BLUP</strong> for data to-be-used as validation values in cross-validation, and a column of weights (<strong>WT</strong>) that could just be set to 1 if users aren’t doing weighted anlaysis.</li>
</ul></li>
<li><p><input type="checkbox" disabled>
What would make the package BrAPI compliant?</p></li>
<li><p><input type="checkbox" disabled>
SNP marker ID naming conventions are probably too strict; add flexibility / robustness</p></li>
<li><p><input type="checkbox" disabled>
More flexibility for the <code><a href="../reference/crosses2predict.html">crosses2predict()</a></code> function: reciprocal crosses? don’t include selfs?</p></li>
<li><p><input type="checkbox" disabled>
Change <code><a href="../reference/predCrossMeans.html">predCrossMeans()</a></code> so that it can accept either <code>dosages</code> or a <code>haploMat</code>. Currently, <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> requires users to supply both <code>dosages</code> <em>and</em> <code>haploMat</code> . Changing <code><a href="../reference/predCrossMeans.html">predCrossMeans()</a></code> in this way will allow users of <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> to supply only the <code>haploMat</code> if they are predicting both cross means <em>and</em> variances, or only <code>dosages</code> if predicting only means.</p></li>
<li><p><input type="checkbox" disabled><code>modelType="AD_geno"</code> or something like that –&gt; genotypic additive-dominance partition allowing prediction of <span class="math inline"><em>μ</em><sub><em>T</em><em>G</em><em>V</em></sub></span> in addition to <span class="math inline"><em>μ</em><sub><em>B</em><em>V</em></sub></span> but <em>without</em> the genome-wide directional dominance part implemented in <code>modelType="DirDom"</code>.</p></li>
<li><p><input type="checkbox" disabled>
When you set <code>getMarkEffs=TRUE</code> in <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code>, tell the user that the function is erroring because they forgot to input the dosage matrix via <code>dosages=</code>.</p></li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marnin Wolfe.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

