<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genomic predictions with non-additive effects • genomicMateSelectR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Genomic predictions with non-additive effects">
<meta property="og:description" content="genomicMateSelectR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genomicMateSelectR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/non_additive_models.html">Genomic predictions with non-additive effects</a>
    </li>
    <li>
      <a href="../articles/start_here.html">Getting started predicting crosses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wolfemd/genomicMateSelectR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genomic predictions with non-additive effects</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/wolfemd/genomicMateSelectR/blob/HEAD/vignettes/non_additive_models.Rmd" class="external-link"><code>vignettes/non_additive_models.Rmd</code></a></small>
      <div class="hidden name"><code>non_additive_models.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wolfemd/genomicMateSelectR" class="external-link">genomicMateSelectR</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: tibble</span>
<span class="co">#&gt; Loading required package: dplyr</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="co">#&gt; Loading required package: tidyr</span>
<span class="co">#&gt; Loading required package: purrr</span></code></pre></div>
<div class="section level1">
<h1 id="genetic-models-implemented">Genetic models implemented<a class="anchor" aria-label="anchor" href="#genetic-models-implemented"></a>
</h1>
<div class="section level2">
<h2 id="additive-only-models">Additive-only models<a class="anchor" aria-label="anchor" href="#additive-only-models"></a>
</h2>
<p>Previously, in the <a href="articles/start_here.html">quick start vignette</a>, I showed how to use the <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> function with a single, additive genetic effect, by setting <code>modelType="A"</code>. We supplied example marker-effects as input via the <code>snpeffs=</code> argument.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snpeffsA</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Trait</span>,<span class="va">allelesubsnpeff</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span>
<span class="co">#&gt;   Trait  allelesubsnpeff</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait2 <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span></span></code></pre></div>
<p>These marker effects (<code>snpeffsA</code>) were derived from standard genome-wide marker regression (i.e. RR-BLUP, a.k.a. SNP-BLUP).</p>
<p>Consider a diploid population with <em>n</em> individuals genotyped at <em>p</em> biallelic genomic loci.</p>
<p>The standard genomic mixed-model is of the form:</p>
<p><span class="math display">\[\boldsymbol{y} =\boldsymbol{X\beta} +\boldsymbol{Z\alpha} +\boldsymbol{\epsilon}\]</span></p>
<p>In this linear model, the <span class="math inline">\(n \times 1\)</span> vector of phenotypic observations, <span class="math inline">\(\boldsymbol{y}\)</span> is modeled according to a combination of genetic and non-genetic effects. Fixed experimental design-related effects estimates are given by <span class="math inline">\(\boldsymbol{\beta}\)</span> and its corresponding incidence matrix <span class="math inline">\(\boldsymbol{X}\)</span> (<span class="math inline">\([ n \times N_{fixed}]\)</span>) where <span class="math inline">\(N_{fixed}\)</span> is the number of fixed factors. The elements of the <span class="math inline">\([n \times p]\)</span> matrix <span class="math inline">\(\boldsymbol{Z}\)</span> contains column-centered marker genotypes:</p>
<p><span class="math display">\[\begin{matrix} z_{ij} = \begin{cases} 
2-2p_{j} &amp; A_1A_1\\
1-2p_{j} &amp; A_1A_2\\
0-2p_{j} &amp; A_2A_2
\end{cases}\end{matrix}\]</span></p>
<p>Here, <span class="math inline">\(p_j\)</span> and <span class="math inline">\(q_j\)</span> are the population frequencies for the <span class="math inline">\(A_1\)</span> and <span class="math inline">\(A_2\)</span> alleles, respectively for the <em>j</em>th SNP marker.</p>
<p>The marker-effects (<span class="math inline">\(\boldsymbol{\alpha}\)</span>) are fitted as independent and identically distributed (<em>i.i.d.</em>) random-effects with mean zero and effects-variance <span class="math inline">\(\sigma^2_{\alpha}\)</span>, <em>i.e.</em> <span class="math inline">\(\alpha \sim N(0,\sigma^2_{\alpha})\)</span>.</p>
<p>As specified <span class="math inline">\(\alpha\)</span> represent allele-substitution effects.</p>
<p>Genomic estimated breeding values (GEBV), which are individual-level (rather than marker-level) selection criteria, can therefore be predicted as <span class="math inline">\(\boldsymbol{\hat{g}}_{BV} =\boldsymbol{Z\hat{\alpha}}\)</span> based on this model.</p>
</div>
<div class="section level2">
<h2 id="models-including-dominance">Models including dominance<a class="anchor" aria-label="anchor" href="#models-including-dominance"></a>
</h2>
<p>We focus in this package, thus far, on dominance effects. Epistasis could be added, in principle.</p>
<p>It is worth reviewing some subtle but important differences between statistical models for partitioning the effect of a locus on a phenotype into “additive” and “dominance” components. <code>genomicMateSelectR</code>’s higher-level functions (<code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> and <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> are careful about this; I want to make the user aware.</p>
<p>To really understand the matter, there is a strong, recent literature on non-additive effects in genomic prediction, I recommend reviewing: <span class="citation">Vitezica, Varona, and Legarra (2013)</span>; <span class="citation">Varona et al. (2018)</span>; <span class="citation">Xiang et al. (2016)</span>; <span class="citation">Wolfe et al. (2021)</span>.</p>
<p>There are two partitions of additive and dominance effects used by <code>genomicMateSelectR</code>, one for <code>modelType="AD"</code>, the other relevant for <code>modelType="DirDom"</code>.</p>
<div class="section level3">
<h3 id="classical-statistical">Classical (statistical)<a class="anchor" aria-label="anchor" href="#classical-statistical"></a>
</h3>
<p>Vitezica <em>et al.</em> <span class="citation">(2013)</span> showed the “classical” model specification, which partitions the total genotypic effect into <strong>allele substitution</strong> and <strong>dominance deviations</strong>.</p>
<p><span class="math display">\[\boldsymbol{y} =\boldsymbol{X\beta} +\boldsymbol{Z\alpha} +\boldsymbol{Wd} +\boldsymbol{\epsilon}\]</span></p>
<p><span class="math inline">\(\boldsymbol{Z\alpha}\)</span> are as in <code>modelType="A"</code>, column-centered dosages and allele substitution effects.</p>
<p>The new matrix of column-centered dominance codings, <span class="math inline">\(\boldsymbol{W}\)</span> contains:</p>
<p><span class="math display">\[\begin{matrix} w_{ij} = \begin{cases}
-2q_j^2 &amp; A_1A_1\\
2p_jq_j &amp; A_1A_2\\
-2p_j^2 &amp; A_2A_2
\end{cases}\end{matrix}\]</span></p>
<p>The dominance deviation marker-effects (<span class="math inline">\(\boldsymbol{d}\)</span>) are also fitted as <em>i.i.d.</em> random, <span class="math inline">\(d \sim N(0,\sigma^2_{d})\)</span>. The corresponding individual-level, genome-wide prediction, the genomic estimated dominance deviation (<strong>GEDD</strong>) is therefore <span class="math inline">\(\boldsymbol{\hat{g}}_{DD} =\boldsymbol{W\hat{d}}\)</span> and we can predict the total merit or performance (rather than just the breeding value) of an individual as <strong>GETGV = GEBV + GEDD</strong>.</p>
<p>This model is implemented by <code>genomicMateSelectR</code> when the user chooses <code>modelType="AD"</code>.</p>
</div>
<div class="section level3">
<h3 id="genotypic-biological">Genotypic (biological)<a class="anchor" aria-label="anchor" href="#genotypic-biological"></a>
</h3>
<p>There is a subtly different model models biological dominance effects, by using a 0, 1, 0 coding of genotypes <span class="math inline">\(A_1A_1\)</span>, <span class="math inline">\(A_1A_2\)</span> and <span class="math inline">\(A_2A_2\)</span>, respectively.</p>
<p><span class="math display">\[\boldsymbol{y} = \boldsymbol{X\beta} + \boldsymbol{Za} + \boldsymbol{\Gamma} \boldsymbol{d}^* + \boldsymbol{\epsilon}\]</span> The dominance coding in the matrix <span class="math inline">\(\boldsymbol{\Gamma}\)</span> is</p>
<p><span class="math display">\[\begin{matrix} \gamma_{ij} = \begin{cases} 
(0-2p_{j}q_{j}) &amp; A_1A_1\\
(1-2p_{j}q_{j}) &amp; A_1A_2\\
(0-2p_{j}q_{j}) &amp; A_2A_2
\end{cases} \end{matrix}\]</span></p>
<p>The coding for the additive term (<span class="math inline">\(\boldsymbol{Z}\)</span>) is the same as all previous models. However, the resulting set of <em>genotypic additive-effects</em> (<span class="math inline">\(\boldsymbol{a}\)</span>) and the corresponding <em>genotypic dominance-effects</em> <span class="math inline">\(\boldsymbol{d^*}\)</span> <strong><em>are not</em></strong> allele substitution or dominance deviation effects. In <code>genomicMateSelectR</code> the genomic estimated values predicted by this model are referred to as: <strong>GETGV = GEadd + GEdom</strong>.</p>
<p>Allele substitution effects (<span class="math inline">\(\alpha\)</span>) and subsequently <strong>GEBV</strong> can be recovered from this model via:</p>
<p><span class="math display">\[\boldsymbol{\alpha} =\boldsymbol{a} +\boldsymbol{d}(\boldsymbol{q}-\boldsymbol{p})\]</span></p>
<p>However, this is inefficient. In fact, currently, <code>genomicMateSelectR</code> does not both to implement the “genotypic” model as it’s own <code>modelType=</code>; but its on the ‘possibly to-do’ list.</p>
<p>Instead the “genotypic” model is used with small extension in an implementation of Xiang et al. (-<span class="citation">Xiang et al. (2016)</span> )’s model with directional dominance.</p>
</div>
<div class="section level3">
<h3 id="directional-dominance">Directional dominance<a class="anchor" aria-label="anchor" href="#directional-dominance"></a>
</h3>
<p><code>genomicMateSelectR</code> includes a third <code>modelType="DirDom"</code> .</p>
<p>Genomic mixed-model with a genome-wide directional dominance term:</p>
<p><span class="math display">\[\boldsymbol{y} = \boldsymbol{X\beta} + \boldsymbol{f}b + \boldsymbol{Za} + \boldsymbol{\Gamma} \boldsymbol{d}^* + \boldsymbol{\epsilon}\]</span></p>
<p>This is the “genotypic” model with an additive fixed-effect covariate, <span class="math inline">\(\textbf{f}\)</span>, a vector of the individual-level genome-wide proportion of loci that are homozygous. The fixed-effect estimate <span class="math inline">\(\boldsymbol{b}\)</span> is the estimated linear effect of overall homozygosity, interpreted as inbreeding depression or heterosis depending on its direction relative to each trait.</p>
<p><code>genomicMateSelectR</code> actually incorporates the directional-dominance effects into predictions of <strong>GEBV</strong>, <strong>GETGV</strong> as well as cross-means and variances by dividing <span class="math inline">\(b\)</span> by the number of effects (<span class="math inline">\(p\)</span>) and subtracting that value from the vector of dominance effects, to get <span class="math inline">\(\boldsymbol{d} = \boldsymbol{d}^*-\frac{b}{p}\)</span>. The notation here isn’t perfect, because <span class="math inline">\(\boldsymbol{d}\)</span> <em>is not</em> the classical dominance deviation effect, which I notated the same. Instead, here <span class="math inline">\(\boldsymbol{d}\)</span> is the vector of genotypic dominance effects with a potentially non-zero mean dominance value; the directional-dominance term is factored in.</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="implementation-using-rungenomicpredictions-in-genomicmateselectr">Implementation using <code>runGenomicPredictions()</code> in <code>genomicMateSelectR</code><a class="anchor" aria-label="anchor" href="#implementation-using-rungenomicpredictions-in-genomicmateselectr"></a>
</h1>
<p>There are many options for fitting the mixed-models described above, using software both in (and out) of R.</p>
<p>IMHO, the best R package, for the frequentist, remains to be the <code>sommer</code> R package <span class="citation">Covarrubias-Pazaran (2016)</span>. <code>sommer</code> is <a href="https://cran.r-project.org/web/packages/sommer/index.html" class="external-link">available on CRAN</a> and has a number of excellent vignettes to get new users started. <code>sommer</code>’s primary mixed-model solver <code>mmer()</code> is almost as flexible as <code>asreml</code>, but free and open-source. Downside is, it is slower and more memory intensive.</p>
<p><code>genomicMateSelectR</code>’s genomic prediction function uses <code><a href="https://rdrr.io/pkg/sommer/man/mmer.html" class="external-link">sommer::mmer()</a></code> under the hood. The function <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> will implement, across potentially multiple traits, univariate genomic predictions, with error-variances weighted according to user-specific weights.</p>
<div class="section level2">
<h2 id="input-blups-or-other-phenotypes">Input: BLUPs or other phenotypes<a class="anchor" aria-label="anchor" href="#input-blups-or-other-phenotypes"></a>
</h2>
<p><code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> requires phenotypes (response data) for genomic predictions via the argument <code>blups=</code>.</p>
<p>An example input <code>blups</code> is shown below:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span>
<span class="co">#&gt;   Trait  TrainingData      </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>            </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 <span style="color: #949494;">&lt;tibble [300 × 4]&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait2 <span style="color: #949494;">&lt;tibble [300 × 4]&gt;</span></span></code></pre></div>
<p>The tibble <code>blups</code> has one row per trait, one column <strong>Trait</strong> with chr-string of trait labels, second column is called <strong>TrainingData</strong> and is list-type with each element containing a <code>tibble</code>. For example:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span><span class="op">$</span><span class="va">TrainingData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">str</span>
<span class="co">#&gt; tibble [300 × 4] (S3: tbl_df/tbl/data.frame)</span>
<span class="co">#&gt;  $ GID    : chr [1:300] "1" "2" "3" "4" ...</span>
<span class="co">#&gt;  $ drgBLUP: num [1:300] -0.1 -0.99 -0.283 -1.567 -0.949 ...</span>
<span class="co">#&gt;  $ BLUP   : num [1:300] -0.1 -0.99 -0.283 -1.567 -0.949 ...</span>
<span class="co">#&gt;  $ WT     : num [1:300] 1 1 1 1 1 1 1 1 1 1 ...</span></code></pre></div>
<p>Four columns, all required in each tibble inside <code>blups$TrainingData</code>:</p>
<ul>
<li>
<code>GID</code>: genotype ID, chr-string</li>
<li>
<code>drgBLUP</code>: the primary response variable. Stands for de-regressed BLUPs, but users can put any single-value-per-GID that they chose as phenotypic training data.</li>
<li>
<code>BLUP</code>: used by the cross-validation functions (<code><a href="../reference/runCrossVal.html">runCrossVal()</a></code> and <code><a href="../reference/runParentWiseCrossVal.html">runParentWiseCrossVal()</a></code> as a slot for validation data. Typically I use the not-deregressed BLUPs. Users can simply have the columns <code>drgBLUP==BLUP</code> if they wish.</li>
<li>
<code>WT</code>: Weights for the error-variance. Used directly as input for <code>sommer::mmer(weights = )</code> argument. Users can simply input a column of 1’s for an un-weighted prediction.</li>
</ul>
<p>Typically, I implement de-regressed BLUPs as</p>
<p><span class="math display">\[drgBLUP = \frac{BLUP}{REL}\]</span></p>
<p>where <span class="math inline">\(REL\)</span> is the reliability of the BLUP, <span class="math inline">\(REL=1-\frac{PEV}{\sigma^2_g}\)</span>. <span class="math inline">\(\sigma^2_g\)</span> as estimated in the mixed-model.</p>
<p>and the weights as</p>
<p><span class="math inline">\(WT = \frac{1-H^2}{(0.1+\frac{1-REL}{REL})*H^2}\)</span>, according to <span class="citation">Garrick, Taylor, and Fernando (2009)</span>.</p>
<p>As indicated above, users are free to supply alternative weights or use an unweighted analysis.</p>
<p>In the future, I can make this more flexible. Users wanting a single-stage or any other flavor of approach for obtaining marker effects and genomic BLUPs can do so and bypass <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> and format their marker-effects for <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="input-genomic-relationship-matrices">Input: genomic relationship matrices<a class="anchor" aria-label="anchor" href="#input-genomic-relationship-matrices"></a>
</h2>
<p>Each of the models discussed above is a genome-wide marker regression; a RR-BLUP, aka SNP-BLUP model. Each model has an exact equivalent genomic-BLUP (G-BLUP model) formed by constructing a genomic relationship matrix (<strong>GRM</strong>) from column-centered additive and dominance coding matrices described above <span class="citation">(Vitezica, Varona, and Legarra 2013; Varona et al. 2018)</span>. In fact, <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> fits GBLUP-style models using user-supplied <strong>GRM</strong>s via the argument <code>grms=</code>. By default, the argument <code>getMarkEffs=FALSE</code> is set, a GBLUP model is fit, and genomic BLUP values, e.g. <strong>GEBV</strong> and <strong>GEDD</strong> are returned to the user without marker-effects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">doseMat</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">str</span>
<span class="co">#&gt;  int [1:300, 1:100] 1 2 2 1 1 0 1 1 1 2 ...</span>
<span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   ..$ : chr [1:300] "1" "2" "3" "4" ...</span>
<span class="co">#&gt;   ..$ : chr [1:100] "1_2" "1_3" "1_6" "1_7" ...</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">doseMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span>
<span class="co">#&gt;   1_2 1_3 1_6 1_7 1_8 1_11 1_14 1_17</span>
<span class="co">#&gt; 1   1   0   2   0   0    0    1    1</span>
<span class="co">#&gt; 2   2   0   1   1   1    0    0    1</span>
<span class="co">#&gt; 3   2   0   2   0   0    0    1    1</span></code></pre></div>
<p>With the <code>doseMat</code> we can easily construct the <strong>GRMs</strong> we need using the <code><a href="../reference/kinship.html">kinship()</a></code> function built-into <code>genomicMateSelectR</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">A</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">doseMat</span>,type <span class="op">=</span> <span class="st">"add"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span>
<span class="co">#&gt; [1] 300 300</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">A</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;           1           2           3           4           5</span>
<span class="co">#&gt; 1 1.6929383  0.18936973  0.39372344  0.26763285  0.27315593</span>
<span class="co">#&gt; 2 0.1893697  1.01254266  0.12403295  0.03319602 -0.06704188</span>
<span class="co">#&gt; 3 0.3937234  0.12403295  1.28023545 -0.07973320  0.06680451</span>
<span class="co">#&gt; 4 0.2676329  0.03319602 -0.07973320  1.13381525 -0.09453974</span>
<span class="co">#&gt; 5 0.2731559 -0.06704188  0.06680451 -0.09453974  1.21536872</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt;  0.4602  0.7735  0.9456  0.9670  1.1120  1.6929</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-an-additive-model-no-marker-effects">Fit an additive model, no marker effects<a class="anchor" aria-label="anchor" href="#fit-an-additive-model-no-marker-effects"></a>
</h2>
<p>As is we have the needed input to <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code>.</p>
<p>Kinship matrices should be formatted as a named-list of matrices. Depending on the model, either only an additive relationship matrix labeled “A,” or an additive <em>and</em> a dominance matrix (labeled “D”).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grms</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>A<span class="op">=</span><span class="va">A</span><span class="op">)</span>
<span class="va">grms</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">str</span>
<span class="co">#&gt; List of 1</span>
<span class="co">#&gt;  $ A: num [1:300, 1:300] 1.693 0.189 0.394 0.268 0.273 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:300] "1" "2" "3" "4" ...</span>
<span class="co">#&gt;   .. ..$ : chr [1:300] "1" "2" "3" "4" ...</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SIwts</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.75</span>,<span class="fl">0.25</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">`names&lt;-`</span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Trait1"</span>,<span class="st">"Trait2"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gpredsA</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runGenomicPredictions.html">runGenomicPredictions</a></span><span class="op">(</span>modelType <span class="op">=</span> <span class="st">"A"</span>, 
                               selInd <span class="op">=</span> <span class="cn">T</span>, SIwts <span class="op">=</span> <span class="va">SIwts</span>,
                               blups <span class="op">=</span> <span class="va">blups</span>, grms <span class="op">=</span> <span class="va">grms</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: furrr</span>
<span class="co">#&gt; Loading required package: future</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -101.534   11:58:41      0           0</span>
<span class="co">#&gt;     2      -98.2273   11:58:41      0           0</span>
<span class="co">#&gt;     3      -97.4802   11:58:41      0           0</span>
<span class="co">#&gt;     4      -97.3541   11:58:41      0           0</span>
<span class="co">#&gt;     5      -97.3447   11:58:41      0           0</span>
<span class="co">#&gt;     6      -97.344   11:58:41      0           0</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -102.208   11:58:41      0           0</span>
<span class="co">#&gt;     2      -92.4952   11:58:41      0           0</span>
<span class="co">#&gt;     3      -90.2181   11:58:42      1           0</span>
<span class="co">#&gt;     4      -89.889   11:58:42      1           0</span>
<span class="co">#&gt;     5      -89.8724   11:58:42      1           0</span>
<span class="co">#&gt;     6      -89.8715   11:58:42      1           0</span></code></pre></div>
<p>Here’s what <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> returns by default.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsA</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span>
<span class="co">#&gt;   gblups             genomicPredOut  </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>          </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> <span style="color: #949494;">&lt;tibble [300 × 5]&gt;</span> <span style="color: #949494;">&lt;tibble [2 × 4]&gt;</span></span></code></pre></div>
<p>The first element <strong>gblups</strong> is a tibble with genomic BLUPs (GEBV in this case) for each trait, and since we supplied <code>SIwts</code>, the GEBV computed on the selection index, labeled <strong>SELIND</strong>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsA</span><span class="op">$</span><span class="va">gblups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 300 × 5</span></span>
<span class="co">#&gt;    GID   predOf SELIND Trait1  Trait2</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1     GEBV   -<span style="color: #BB0000;">0.994</span> -<span style="color: #BB0000;">1.48</span>   0.454 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 2     GEBV   -<span style="color: #BB0000;">0.428</span> -<span style="color: #BB0000;">1.07</span>   1.49  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 3     GEBV   -<span style="color: #BB0000;">0.474</span> -<span style="color: #BB0000;">0.589</span> -<span style="color: #BB0000;">0.130</span> </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 4     GEBV   -<span style="color: #BB0000;">0.735</span> -<span style="color: #BB0000;">0.741</span> -<span style="color: #BB0000;">0.719</span> </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 5     GEBV   -<span style="color: #BB0000;">0.432</span> -<span style="color: #BB0000;">0.561</span> -<span style="color: #BB0000;">0.044</span><span style="color: #BB0000; text-decoration: underline;">5</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 6     GEBV   -<span style="color: #BB0000;">0.731</span> -<span style="color: #BB0000;">1.13</span>   0.456 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 7     GEBV   -<span style="color: #BB0000;">1.14</span>  -<span style="color: #BB0000;">1.57</span>   0.163 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 8     GEBV    0.444  0.422  0.511 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 9     GEBV   -<span style="color: #BB0000;">0.654</span> -<span style="color: #BB0000;">1.42</span>   1.64  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 10    GEBV   -<span style="color: #BB0000;">0.656</span> -<span style="color: #BB0000;">0.461</span> -<span style="color: #BB0000;">1.24</span>  </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 290 more rows</span></span></code></pre></div>
<p>The other element of the output (<strong>genomicPredOut</strong>) contains <code>data.frame</code>s of the fixed-effect estimates (<strong>fixeffs</strong>) and variance component estimates (<strong>varcomps</strong>) from the model. The <strong>gblups</strong> column here just has a more raw version of the GEBV.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsA</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 4</span></span>
<span class="co">#&gt;   Trait  gblups             varcomps     fixeffs     </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 <span style="color: #949494;">&lt;tibble [300 × 2]&gt;</span> <span style="color: #949494;">&lt;df [2 × 4]&gt;</span> <span style="color: #949494;">&lt;df [1 × 5]&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait2 <span style="color: #949494;">&lt;tibble [300 × 2]&gt;</span> <span style="color: #949494;">&lt;df [2 × 4]&gt;</span> <span style="color: #949494;">&lt;df [1 × 5]&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="returning-marker-effects-with-getmarkeffstrue">Returning marker effects with <code>getMarkEffs=TRUE</code><a class="anchor" aria-label="anchor" href="#returning-marker-effects-with-getmarkeffstrue"></a>
</h2>
<p>Setting <code>getMarkEffs=TRUE</code> in <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> will generate RR-BLUP solutions for SNP-effects.</p>
<p>The helper function <code><a href="../reference/backsolveSNPeff.html">backsolveSNPeff()</a></code> is used.</p>
<p><span class="math display">\[Z^T(ZZ^T)^{-1}g\]</span></p>
<p>Where $ are centered allele dosages or dominance genotype codings, and  are the corresponding GBLUP solutions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsA</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runGenomicPredictions.html">runGenomicPredictions</a></span><span class="op">(</span>modelType <span class="op">=</span> <span class="st">"A"</span>, 
                               selInd <span class="op">=</span> <span class="cn">T</span>, SIwts <span class="op">=</span> <span class="va">SIwts</span>,
                               getMarkEffs <span class="op">=</span> <span class="cn">TRUE</span>,dosages <span class="op">=</span> <span class="va">doseMat</span>,
                               blups <span class="op">=</span> <span class="va">blups</span>, grms <span class="op">=</span> <span class="va">grms</span><span class="op">)</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -101.534   11:58:42      0           0</span>
<span class="co">#&gt;     2      -98.2273   11:58:42      0           0</span>
<span class="co">#&gt;     3      -97.4802   11:58:43      1           0</span>
<span class="co">#&gt;     4      -97.3541   11:58:43      1           0</span>
<span class="co">#&gt;     5      -97.3447   11:58:43      1           0</span>
<span class="co">#&gt;     6      -97.344   11:58:43      1           0</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -102.208   11:58:43      0           0</span>
<span class="co">#&gt;     2      -92.4952   11:58:43      0           0</span>
<span class="co">#&gt;     3      -90.2181   11:58:43      0           0</span>
<span class="co">#&gt;     4      -89.889   11:58:43      0           0</span>
<span class="co">#&gt;     5      -89.8724   11:58:43      0           0</span>
<span class="co">#&gt;     6      -89.8715   11:58:43      0           0</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsA</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span>
<span class="co">#&gt;   gblups             genomicPredOut  </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>          </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> <span style="color: #949494;">&lt;tibble [300 × 5]&gt;</span> <span style="color: #949494;">&lt;tibble [2 × 5]&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsA</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 5</span></span>
<span class="co">#&gt;   Trait  gblups             varcomps     fixeffs      allelesubsnpeff</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 <span style="color: #949494;">&lt;tibble [300 × 2]&gt;</span> <span style="color: #949494;">&lt;df [2 × 4]&gt;</span> <span style="color: #949494;">&lt;df [1 × 5]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait2 <span style="color: #949494;">&lt;tibble [300 × 2]&gt;</span> <span style="color: #949494;">&lt;df [2 × 4]&gt;</span> <span style="color: #949494;">&lt;df [1 × 5]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span></span></code></pre></div>
<p>Allele-substitution effects matrix for each trait is now in the list-column <strong>allelesubsnpeff</strong>.</p>
</div>
<div class="section level2">
<h2 id="run-non-additive-genomic-predictions">Run non-additive genomic predictions<a class="anchor" aria-label="anchor" href="#run-non-additive-genomic-predictions"></a>
</h2>
<div class="section level3">
<h3 id="modeltypead">
<code>modelType="AD"</code><a class="anchor" aria-label="anchor" href="#modeltypead"></a>
</h3>
<p>As mentioned above, <code>modelType="AD"</code> effects an additive-plus-dominance model using the “classical” partition. Predictions of GBLUPs (SNP-BLUPs / marker-effects) will correspond to <strong>GEBV</strong> (<strong>allele substitution effects</strong>).</p>
<p>We need first to add the correct <strong>dominance relationship matrix</strong>, constructed with <code><a href="../reference/kinship.html">kinship()</a></code> to <code>grms</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">D</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">doseMat</span>,type <span class="op">=</span> <span class="st">"domClassic"</span><span class="op">)</span>
<span class="va">grms</span><span class="op">[[</span><span class="st">"D"</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;-</span><span class="va">D</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsAD</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runGenomicPredictions.html">runGenomicPredictions</a></span><span class="op">(</span>modelType <span class="op">=</span> <span class="st">"AD"</span>, 
                                selInd <span class="op">=</span> <span class="cn">T</span>, SIwts <span class="op">=</span> <span class="va">SIwts</span>,
                                getMarkEffs <span class="op">=</span> <span class="cn">TRUE</span>,dosages <span class="op">=</span> <span class="va">doseMat</span>,
                                blups <span class="op">=</span> <span class="va">blups</span>, grms <span class="op">=</span> <span class="va">grms</span><span class="op">)</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -100.583   11:58:44      0           0</span>
<span class="co">#&gt;     2      -96.4141   11:58:44      0           0</span>
<span class="co">#&gt;     3      -95.5178   11:58:44      0           0</span>
<span class="co">#&gt;     4      -95.3651   11:58:44      0           0</span>
<span class="co">#&gt;     5      -95.353   11:58:44      0           0</span>
<span class="co">#&gt;     6      -95.3518   11:58:45      1           0</span>
<span class="co">#&gt;     7      -95.3517   11:58:45      1           0</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -98.116   11:58:45      0           0</span>
<span class="co">#&gt;     2      -84.3225   11:58:45      0           0</span>
<span class="co">#&gt;     3      -81.2501   11:58:45      0           0</span>
<span class="co">#&gt;     4      -80.7642   11:58:45      0           0</span>
<span class="co">#&gt;     5      -80.7346   11:58:45      0           0</span>
<span class="co">#&gt;     6      -80.7326   11:58:46      1           0</span>
<span class="co">#&gt;     7      -80.7325   11:58:46      1           0</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsAD</span><span class="op">$</span><span class="va">gblups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span>
<span class="co">#&gt;   GID   predOf SELIND Trait1 Trait2</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1     GEBV   -<span style="color: #BB0000;">0.922</span> -<span style="color: #BB0000;">1.48</span>   0.753</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1     GETGV  -<span style="color: #BB0000;">0.922</span> -<span style="color: #BB0000;">1.48</span>   0.753</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 2     GEBV   -<span style="color: #BB0000;">0.362</span> -<span style="color: #BB0000;">1.09</span>   1.82 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 2     GETGV  -<span style="color: #BB0000;">0.362</span> -<span style="color: #BB0000;">1.09</span>   1.82 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 3     GEBV   -<span style="color: #BB0000;">0.455</span> -<span style="color: #BB0000;">0.558</span> -<span style="color: #BB0000;">0.146</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 3     GETGV  -<span style="color: #BB0000;">0.455</span> -<span style="color: #BB0000;">0.558</span> -<span style="color: #BB0000;">0.146</span></span></code></pre></div>
<p>Note that for each <strong>GID</strong> there are now two rows, distinguished by the <strong>predOf</strong> column, as either <strong>GEBV</strong> or <strong>GETGV</strong> (the <strong>GEBV</strong> + <strong>GEDD</strong>).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsAD</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 6</span></span>
<span class="co">#&gt;   Trait  gblups             varcomps     fixeffs   allelesubsnpeff domdevsnpeff </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>       </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 <span style="color: #949494;">&lt;tibble [300 × 4]&gt;</span> <span style="color: #949494;">&lt;df [3 × 4]&gt;</span> <span style="color: #949494;">&lt;df [1 ×</span>… <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × </span>…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait2 <span style="color: #949494;">&lt;tibble [300 × 4]&gt;</span> <span style="color: #949494;">&lt;df [3 × 4]&gt;</span> <span style="color: #949494;">&lt;df [1 ×</span>… <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × </span>…</span></code></pre></div>
<p>Now two sets of marker effects are reported: <strong>allelesubsnpeff</strong> and <strong>domdevsnpeff</strong>.</p>
</div>
<div class="section level3">
<h3 id="modeltypedirdom">
<code>modelType="DirDom"</code><a class="anchor" aria-label="anchor" href="#modeltypedirdom"></a>
</h3>
<p>The last model, implements an additive-plus-directional dominance model as detailed above. The “DirDom” model adds the potential for a genome-wide inbreeding depression (or heterosis effect).</p>
<p>To do this correctly, we need to switch to the “genotyped” parameterization of dominance (see above, and <span class="citation">Vitezica, Varona, and Legarra (2013)</span>.</p>
<p>Simply change the <code>type=</code> argument to “domGenotypic” in the <code><a href="../reference/kinship.html">kinship()</a></code> function to create a <strong>genotypic dominance relationship matrix</strong></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">D</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">doseMat</span>,type <span class="op">=</span> <span class="st">"domGenotypic"</span><span class="op">)</span>
<span class="co"># Replace the previous "D" matrix, but don't change the name</span>
<span class="co"># grms list input must have names "A" or "D".</span>
<span class="va">grms</span><span class="op">[[</span><span class="st">"D"</span><span class="op">]</span><span class="op">]</span><span class="op">&lt;-</span><span class="va">D</span></code></pre></div>
<p>That’s the only different input that is needed.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsDirDom</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runGenomicPredictions.html">runGenomicPredictions</a></span><span class="op">(</span>modelType <span class="op">=</span> <span class="st">"DirDom"</span>, 
                                    selInd <span class="op">=</span> <span class="cn">T</span>, SIwts <span class="op">=</span> <span class="va">SIwts</span>,
                                    getMarkEffs <span class="op">=</span> <span class="cn">TRUE</span>,dosages <span class="op">=</span> <span class="va">doseMat</span>,
                                    blups <span class="op">=</span> <span class="va">blups</span>, grms <span class="op">=</span> <span class="va">grms</span><span class="op">)</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -99.0157   11:58:46      0           0</span>
<span class="co">#&gt;     2      -96.3744   11:58:46      0           0</span>
<span class="co">#&gt;     3      -95.732   11:58:47      1           0</span>
<span class="co">#&gt;     4      -95.6263   11:58:47      1           0</span>
<span class="co">#&gt;     5      -95.6193   11:58:47      1           0</span>
<span class="co">#&gt;     6      -95.6188   11:58:47      1           0</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -90.9376   11:58:47      0           0</span>
<span class="co">#&gt;     2      -81.2191   11:58:47      0           0</span>
<span class="co">#&gt;     3      -78.5672   11:58:47      0           0</span>
<span class="co">#&gt;     4      -78.0987   11:58:47      0           0</span>
<span class="co">#&gt;     5      -78.0659   11:58:48      1           0</span>
<span class="co">#&gt;     6      -78.0635   11:58:48      1           0</span>
<span class="co">#&gt;     7      -78.0633   11:58:48      1           0</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsDirDom</span><span class="op">$</span><span class="va">gblups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span>
<span class="co">#&gt;   GID   predOf   SELIND Trait1  Trait2</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1     GEBV     0.043<span style="text-decoration: underline;">3</span>  32.9  -<span style="color: #BB0000;">98.4</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1     GETGV    2.88     3.54   0.878</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 2     GEBV     4.51   -<span style="color: #BB0000;">23.5</span>   88.4  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 2     GETGV   -<span style="color: #BB0000;">4.78</span>    -<span style="color: #BB0000;">7.38</span>   3.00 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 3     GEBV   -<span style="color: #BB0000;">11.8</span>    -<span style="color: #BB0000;">33.5</span>   53.6  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 3     GETGV   -<span style="color: #BB0000;">1.59</span>    -<span style="color: #BB0000;">1.99</span>  -<span style="color: #BB0000;">0.403</span></span></code></pre></div>
<p>Note: no difference in structure of output <strong>gblups</strong>. For this model, rather than using the genomic BLUPs computed by the initial GBLUP model, <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> first backsolves for the SNP-effects, then adds the genome-wide directional dominance effect. Because of this approach, we can also recover the allele substitution effects by doing <span class="math inline">\(\alpha = a + d(q-p)\)</span>was added to the SNP-effects and then genomic BLUPs were calculated to get GEBV and GETGV that include this effect.</p>
<p>The estimated genome-wide directional dominance effect itself is available in the <strong>fixeffs</strong> column:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsDirDom</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">fixeffs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt;     Trait      Effect    Estimate Std.Error     t.value</span>
<span class="co">#&gt; 1 drgBLUP (Intercept)  0.91088794  1.083581  0.84062719</span>
<span class="co">#&gt; 2 drgBLUP           f -0.08215501  1.521228 -0.05400573</span></code></pre></div>
<p>The SNP effects themselves:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsDirDom</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 8</span></span>
<span class="co">#&gt;   Trait  gblups    varcomps  fixeffs   allelesubsnpeff addsnpeff  domstar_snpeff</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>        </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 <span style="color: #949494;">&lt;tibble </span>… <span style="color: #949494;">&lt;df [3 ×</span>… <span style="color: #949494;">&lt;df [2 ×</span>… <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100</span>… <span style="color: #949494;">&lt;dbl [100 × 1</span>…</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait2 <span style="color: #949494;">&lt;tibble </span>… <span style="color: #949494;">&lt;df [3 ×</span>… <span style="color: #949494;">&lt;df [2 ×</span>… <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100</span>… <span style="color: #949494;">&lt;dbl [100 × 1</span>…</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 1 more variable: domsnpeff &lt;list&gt;</span></span></code></pre></div>
<p>Marker effects reported by the <strong>DirDom</strong> model:</p>
<ul>
<li>
<code>allelesubsnpeff</code>: allele substition effects with genome-wide directional dominance effect added</li>
<li>
<code>addsnpeff</code>: genotypic additive effect</li>
<li>
<code>domstar_snpeff</code>: genotypic dominance effect</li>
<li>
<code>domsnpeff</code>: genotypic dominance effects with genome-wide directional dominance effect added</li>
</ul>
</div>
</div>
</div>
<div class="section level1">
<h1 id="cross-predictions-with-genomicmateselectr">Cross predictions with <code>genomicMateSelectR</code><a class="anchor" aria-label="anchor" href="#cross-predictions-with-genomicmateselectr"></a>
</h1>
<div class="section level2">
<h2 id="theory">Theory<a class="anchor" aria-label="anchor" href="#theory"></a>
</h2>
<p>See <span class="citation">Wolfe et al. (2021)</span> for full details.</p>
<p>Briefly:</p>
<p><strong>Predict cross mean breeding values</strong></p>
<p><span class="math display">\[\mu_{BV}=\frac{GEBV_{P1}+GEBV_{P2}}{2}\]</span></p>
<p><strong>Predict cross mean total genetic value</strong></p>
<p>(also called <strong>genomic prediction of cross performance</strong> or <strong>GPCP</strong> by <span class="citation">Werner et al. (2020)</span>) <span class="math display">\[\mu_{TGV}=\sum\limits_{k=1}^p a_k(p_{ik} - q_{ik} - y_k) + d_k[2p_{ik}q_{ik} + y_k(p_{ik}-q_{ik})]\]</span></p>
<p>See <span class="citation">(Toro and Varona 2010; Varona et al. 2018)</span>.</p>
<p>Here <span class="math inline">\(p_{ik}\)</span> and <span class="math inline">\(q_{ik}\)</span> are the allele frequencies of the counted (alternative) and the non-counted (reference-genome) allele, respectively, for one of the two parents (indexed by <span class="math inline">\(i\)</span>). The difference in frequency between the parent one (indexed by <span class="math inline">\(i\)</span>) and the parent two (indexed by <span class="math inline">\(j\)</span>) is ,<span class="math inline">\(y_k = p_{ik} - p_{jk}\)</span> and the summation is over the <span class="math inline">\(p\)</span> markers indexed by <span class="math inline">\(k\)</span>. Note that <span class="math inline">\(a_k\)</span> is the average effect and <em>not</em> the allele substitution effect, <span class="math inline">\(\alpha\)</span>.</p>
<p>The genotypic parameterization for dominance should be to get marker effects to predict <span class="math inline">\(\mu_{TGV}\)</span> in this way. In fact, for this reason, <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> only returns <span class="math inline">\(\mu_{TGV}\)</span> predictions when <code>modelType="DirDom"</code>.</p>
<p><strong>Predict the genetic variance among offspring</strong></p>
<p><span class="math display">\[\hat{\sigma}^{2}_{BV} = \boldsymbol{\alpha}^{T}\boldsymbol{D}\boldsymbol{\alpha}\]</span> <span class="math display">\[\hat{\sigma}^{2}_{DD} = \boldsymbol{d}^{T}\boldsymbol{D}^2\boldsymbol{d}\]</span></p>
<p>Where <span class="math inline">\(\boldsymbol{D}^2 =\boldsymbol{D}\odot\boldsymbol{D}\)</span>, <span class="math inline">\(\odot\)</span> indicating element-wise (Hadamard) multiplication of <span class="math inline">\(\boldsymbol{D}\)</span>, having the effect of squaring all elements.</p>
<p><span class="math display">\[\hat{\sigma}^{2}_{TGV} = \hat{\sigma}^{2}_{BV} +\hat{\sigma}^{2}_{DD}\]</span></p>
<p>The <span class="math inline">\(p\times p\)</span> variance-covariance matrix, <span class="math inline">\(\boldsymbol{D}\)</span>, is the expected linkage disequilibrium among full-siblings by considering the expected pairwise recombination frequency and each parent’s haplotype phase.</p>
<p><span class="math display">\[\boldsymbol{D}_{P_1}^{gametes}=(1-2\boldsymbol{c})\odot\boldsymbol{D}_{P_1}^{haplos}\]</span></p>
<p><span class="math display">\[\boldsymbol{D}_{P_2}^{gametes}=(1-2\boldsymbol{c})\odot\boldsymbol{D}_{P_2}^{haplos}\]</span></p>
<p><span class="math display">\[\boldsymbol{D}_{P_1\times P_2}^{genotypes} =\boldsymbol{D}_{P_1}^{gametes} +\boldsymbol{D}_{P_2}^{gametes}\]</span></p>
<p><span class="math inline">\(\boldsymbol{D}_{P_1}^{haplos}\)</span> and <span class="math inline">\(\boldsymbol{D}_{P_2}^{haplos}\)</span> are simply the <span class="math inline">\(p\times p\)</span> covariance matrices associated with each parent’s respective <span class="math inline">\(2\times p\)</span> haplotype matrix (<span class="math inline">\(\boldsymbol{H}_{P_{1} or P_{2}}\)</span>), where elements are 1 if the counted allele is present, 0 otherwise.</p>
<p><span class="math display">\[\boldsymbol{D}^{haplos} =\frac{1}{2}\boldsymbol{H}^T\boldsymbol{H} -\boldsymbol{p}\boldsymbol{p}^T\]</span> Where <span class="math inline">\(\boldsymbol{p}\)</span> is a vector of within-individual, per-SNP allele frequencies.</p>
<p>The <span class="math inline">\(p\times p\)</span> pairwise recombination frequencies matrix is <span class="math inline">\(\boldsymbol{c}\)</span> and can be derived from a genetic map. <span class="math inline">\(\boldsymbol{D}_{P_1}^{gametes}\)</span> and <span class="math inline">\(\boldsymbol{D}_{P_2}^{gametes}\)</span> are the covariance matrices for each parents pool of possible gametes, whose covariances sum to give the expected covariances genotypes in the cross, <span class="math inline">\(\boldsymbol{D}_{P_1\times P_2}^{genotypes}\)</span>. The genetic variances <span class="math inline">\(\hat{\sigma}^{2}_{BV}\)</span> and <span class="math inline">\(\hat{\sigma}^{2}_{DD}\)</span> are thus predicted as above by using <span class="math inline">\(\boldsymbol{D} = \boldsymbol{D}_{P_1\times P_2}^{genotypes}\)</span>.</p>
<p>See <span class="citation">(Lehermeier, Teyssèdre, and Schön 2017)</span>.</p>
<p><strong>Predicting the trait-trait co-variances among offspring</strong></p>
<p>Consider two traits, <span class="math inline">\(T1\)</span> and <span class="math inline">\(T2\)</span></p>
<p>Variance for <span class="math inline">\(T1\)</span>: <span class="math display">\[\sigma^2_{T1}=\boldsymbol{\alpha}_{T1}^{T}\boldsymbol{D}\boldsymbol{\alpha}_{T1}\]</span></p>
<p>Variance for <span class="math inline">\(T2\)</span>: <span class="math display">\[\sigma^2_{T2}=\boldsymbol{\alpha}_{T2}^{T}\boldsymbol{D}\boldsymbol{\alpha}_{T2}\]</span></p>
<p>Covariance between <span class="math inline">\(T1\)</span> and <span class="math inline">\(T2\)</span>: <span class="math display">\[\sigma_{T1,T2}=\boldsymbol{\alpha}_{T1}^{T}\boldsymbol{D}\boldsymbol{\alpha}_{T2}\]</span></p>
<p>Apply to dominance by substituting <span class="math inline">\(\boldsymbol{\alpha}\)</span> with <span class="math inline">\(\boldsymbol{d}\)</span> and squaring elements of <span class="math inline">\(\boldsymbol{D}\)</span>.</p>
<p>See <span class="citation">(Neyhart, Lorenz, and Smith 2019)</span>.</p>
<p><strong>Predict means and variances among offspring for the selection index</strong></p>
<p>Selection index mean: <span class="math display">\[\mu_{SI} =\boldsymbol{w}^T\hat{\boldsymbol{g}}_{BV}\]</span></p>
<p>Selection index variance: <span class="math display">\[\sigma^2_{SI} =\boldsymbol{w}^T\boldsymbol{G}\boldsymbol{w}\]</span></p>
<p>The <span class="math inline">\(n\times T\)</span> matrix <span class="math inline">\(\hat{\boldsymbol{g}}_{BV}\)</span> contains the GEBV for each trait and the <span class="math inline">\(T\times 1\)</span> vector <span class="math inline">\(\boldsymbol{w}\)</span> are the index weights. The <span class="math inline">\(T\times T\)</span> matrix <span class="math inline">\(\boldsymbol{G}\)</span> is the additive (or total) genetic variance-covariance matrix for traits on the index.</p>
<p><span class="math display">\[\boldsymbol{G} = \begin{bmatrix} 
\sigma^2_{Trait_1,Trait_1} &amp; \hdots &amp; \sigma_{Trait_1,Trait_T}\\
\vdots &amp; \ddots &amp; \vdots \\
\sigma_{Trait_1,Trait_T} &amp; \hdots &amp; \sigma^2_{Trait_T,Trait_T}\end{bmatrix}\]</span></p>
<p>See <span class="citation">(Bonk et al. 2016)</span>.</p>
<p><strong>Predict the (selection-index) usefulness of the cross</strong> (<strong>the superior progeny mean</strong>)</p>
<p><span class="math display">\[\boldsymbol{UC}_{SI}=\boldsymbol{\mu}_{SI} +\boldsymbol{i}_{SI}\times\boldsymbol{\sigma}_{SI}\]</span></p>
<p>Distinguish two types of usefulness, concerning breeding value (allele substitution effects) versus total genetic value (total genetic effects, additive + dominance).</p>
<p><span class="math display">\[\hat{UC}_{parent} = \hat{\mu}_{BV}+i\times \hat{\sigma}_{BV}\]</span> <span class="math display">\[\hat{UC}_{variety} = \hat{\mu}_{TGV}+i\times \hat{\sigma}_{TGV}\]</span> For simplicity, dropped the “SI” notation.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>In the <a href="articles/start_here.html">previous vignette</a>, I covered the formatting for inputs to <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> when the <code>modelType="A"</code>. Below, I’ll highlight the features and implementation of the non-additive models.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># recombFreqMat</span>
<span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/genmap2recombfreq.html">genmap2recombfreq</a></span><span class="op">(</span><span class="va">genmap</span>, nChr <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fl">1</span><span class="op">-</span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">recombFreqMat</span><span class="op">)</span>


<span class="co"># Crosses to predict</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span>;
<span class="va">parents</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">snpeffsA</span><span class="op">$</span><span class="va">gblups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">GID</span>, size <span class="op">=</span> <span class="fl">5</span>, replace <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">CrossesToPredict</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/crosses2predict.html">crosses2predict</a></span><span class="op">(</span><span class="va">parents</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="modeltypead-1">`modelType=“AD”<a class="anchor" aria-label="anchor" href="#modeltypead-1"></a>
</h3>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">doseMat</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">haploMat</span><span class="op">)</span>
<span class="va">predVarsAD</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/predictCrosses.html">predictCrosses</a></span><span class="op">(</span>modelType <span class="op">=</span> <span class="st">"AD"</span>,
                           selInd <span class="op">=</span> <span class="cn">TRUE</span>, SIwts <span class="op">=</span> <span class="va">SIwts</span>,
                           CrossesToPredict<span class="op">=</span><span class="va">CrossesToPredict</span>,
                           snpeffs<span class="op">=</span><span class="va">gpredsAD</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                           dosages<span class="op">=</span><span class="va">doseMat</span>,
                           haploMat<span class="op">=</span><span class="va">haploMat</span>,recombFreqMat<span class="op">=</span><span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt; [1] "Predicting cross variance parameters"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.12 mins for 15 crosses"</span>
<span class="co">#&gt; [1] "Predicting cross means"</span>
<span class="co">#&gt; [1] "Computing SELECTION INDEX means and variances."</span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predVarsAD</span><span class="op">$</span><span class="va">tidyPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 9</span></span>
<span class="co">#&gt;   sireID damID Nsegsnps predOf Trait  predMean predVar predSD predUsefulness</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 49     49          22 BV     SELIND    130.     334.   18.3          179. </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 49     49          22 BV     Trait1    203.     640.   25.3          271. </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 49     49          22 BV     Trait2    -<span style="color: #BB0000;">89.2</span>   <span style="text-decoration: underline;">1</span>597.   40.0           17.5</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 49     153         46 BV     SELIND    117.     479.   21.9          175. </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 49     153         46 BV     Trait1    180.    <span style="text-decoration: underline;">1</span>064.   32.6          267. </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 49     153         46 BV     Trait2    -<span style="color: #BB0000;">73.4</span>   <span style="text-decoration: underline;">1</span>815.   42.6           40.3</span></code></pre></div>
<p>For <code>modelType="AD"</code> there should be two rows per cross per trait and in this case the column <strong>Trait</strong> also contains predictions on the selection index (<strong>SELIND</strong>).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predVarsAD</span><span class="op">$</span><span class="va">tidyPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">arrange</span><span class="op">(</span><span class="va">Trait</span>,<span class="va">sireID</span>,<span class="va">damID</span>,<span class="va">predOf</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 9</span></span>
<span class="co">#&gt;   sireID damID Nsegsnps predOf Trait  predMean predVar predSD predUsefulness</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 146    146         34 BV     SELIND     82.1    509.   22.6           142.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 146    146         34 TGV    SELIND     82.1    521.   22.8           143.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 153    146         48 BV     SELIND     92.5    567.   23.8           156.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 153    146         48 TGV    SELIND     92.5    579.   24.1           157.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 153    153         33 BV     SELIND    103.     624.   25.0           170.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 153    153         33 TGV    SELIND    103.     640.   25.3           170.</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predVarsAD</span><span class="op">$</span><span class="va">rawPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">predMeans</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu">arrange</span><span class="op">(</span><span class="va">Trait</span>,<span class="va">sireID</span>,<span class="va">damID</span>,<span class="va">predOf</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span>
<span class="co">#&gt;   Trait  sireID damID sireGEBV damGEBV predOf predMean</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 146    146       110.    110. MeanBV     110.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait1 146    146       110.    110. TGV        110.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Trait1 153    146       156.    110. MeanBV     133.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Trait1 153    146       156.    110. TGV        133.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Trait1 153    153       156.    156. MeanBV     156.</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Trait1 153    153       156.    156. TGV        156.</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predVarsAD</span><span class="op">$</span><span class="va">rawPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">predVars</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu">arrange</span><span class="op">(</span><span class="va">sireID</span>,<span class="va">damID</span>,<span class="va">predOf</span>,<span class="va">Trait1</span>,<span class="va">Trait2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 8 × 8</span></span>
<span class="co">#&gt;   sireID damID Nsegsnps ComputeTime Trait1 Trait2 predOf predVar</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 146    146         34       0.483 Trait1 Trait1 VarBV  <span style="text-decoration: underline;">1</span>232.  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 146    146         34       0.483 Trait1 Trait2 VarBV  -<span style="color: #BB0000;">820.</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 146    146         34       0.483 Trait2 Trait2 VarBV  <span style="text-decoration: underline;">1</span>976.  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 146    146         34       0.483 Trait1 Trait1 VarDD     9.97</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 146    146         34       0.483 Trait1 Trait2 VarDD     9.04</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 146    146         34       0.483 Trait2 Trait2 VarDD    51.3 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">7</span> 153    146         48       0.468 Trait1 Trait1 VarBV  <span style="text-decoration: underline;">1</span>360.  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">8</span> 153    146         48       0.468 Trait1 Trait2 VarBV  -<span style="color: #BB0000;">863.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="modeltypedirdom-1">`modelType=“DirDom”<a class="anchor" aria-label="anchor" href="#modeltypedirdom-1"></a>
</h3>
<p>For the <strong>DirDom</strong> model, not much change is needed.</p>
<p>One note: if you got marker effects using <code>runGenomicPredictions(modelType="DirDom")</code> then the SNP-effects generated, used below (<code>gpredsDirDom$genomicPredOut[[1]]</code>) are all you need. If you want to supply your own marker effects, they input must at a bare minimum contain the columns: <strong>Trait</strong>, <strong>allelesubsnpeff</strong>, <strong>addsnpeff</strong> and <strong>domsnpeff</strong>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gpredsDirDom</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Trait</span>,<span class="va">allelesubsnpeff</span>,<span class="va">addsnpeff</span>,<span class="va">domsnpeff</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 4</span></span>
<span class="co">#&gt;   Trait  allelesubsnpeff addsnpeff       domsnpeff      </span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Trait1 <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Trait2 <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span> <span style="color: #949494;">&lt;dbl [100 × 1]&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predVarsDirDom</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/predictCrosses.html">predictCrosses</a></span><span class="op">(</span>modelType <span class="op">=</span> <span class="st">"DirDom"</span>,
                               selInd <span class="op">=</span> <span class="cn">TRUE</span>, SIwts <span class="op">=</span> <span class="va">SIwts</span>,
                               CrossesToPredict<span class="op">=</span><span class="va">CrossesToPredict</span>,
                               snpeffs<span class="op">=</span><span class="va">gpredsDirDom</span><span class="op">$</span><span class="va">genomicPredOut</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                               dosages<span class="op">=</span><span class="va">doseMat</span>,
                               haploMat<span class="op">=</span><span class="va">haploMat</span>,recombFreqMat<span class="op">=</span><span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt; [1] "Predicting cross variance parameters"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.12 mins for 15 crosses"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.12 mins for 15 crosses"</span>
<span class="co">#&gt; [1] "Predicting cross means"</span>
<span class="co">#&gt; [1] "Computing SELECTION INDEX means and variances."</span></code></pre></div>
<p>Output in same format as <code>modelType="AD"</code>, e.g.:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predVarsDirDom</span><span class="op">$</span><span class="va">tidyPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">str</span>
<span class="co">#&gt; tibble [90 × 9] (S3: tbl_df/tbl/data.frame)</span>
<span class="co">#&gt;  $ sireID        : chr [1:90] "49" "49" "49" "49" ...</span>
<span class="co">#&gt;  $ damID         : chr [1:90] "49" "49" "49" "153" ...</span>
<span class="co">#&gt;  $ Nsegsnps      : int [1:90] 22 22 22 46 46 46 48 48 48 53 ...</span>
<span class="co">#&gt;  $ predOf        : chr [1:90] "BV" "BV" "BV" "BV" ...</span>
<span class="co">#&gt;  $ Trait         : chr [1:90] "SELIND" "Trait1" "Trait2" "SELIND" ...</span>
<span class="co">#&gt;  $ predMean      : num [1:90] 113.2 178.4 -82.2 103.4 157.4 ...</span>
<span class="co">#&gt;  $ predVar       : num [1:90] 264 571 1296 400 954 ...</span>
<span class="co">#&gt;  $ predSD        : num [1:90] 16.3 23.9 36 20 30.9 ...</span>
<span class="co">#&gt;  $ predUsefulness: num [1:90] 157 242 14 157 240 ...</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="additional-predictcrosses-features">Additional <code>predictCrosses()</code> features<a class="anchor" aria-label="anchor" href="#additional-predictcrosses-features"></a>
</h2>
<ul>
<li>Parallel by family (<code>ncores</code>)</li>
<li>Multi-core matrix algrebra (OpenBLAS) usage (if your R compilation is linked to one) (<code>nBLASthreads</code>)</li>
<li>Toggle cross mean and cross variance predictions on-vs-off: <code>predTheMeans</code> and <code>predTheVars</code>.</li>
</ul>
</div>
</div>
<div class="section level1">
<h1 id="next-vignettes">Next vignettes<a class="anchor" aria-label="anchor" href="#next-vignettes"></a>
</h1>
<ul class="task-list">
<li>
<input type="checkbox" disabled>
Cross-validation</li>
<li>
<input type="checkbox" disabled>
Further under the hood: functions <code><a href="../reference/predCrossVars.html">predCrossVars()</a></code> and <code><a href="../reference/predCrossMeans.html">predCrossMeans()</a></code> (used by <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>)</li>
</ul>
</div>
<div class="section level1 unnumbered">
<h1 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bonk2016b" class="csl-entry">
Bonk, Sarah, Manuela Reichelt, Friedrich Teuscher, Dierck Segelke, and Norbert Reinsch. 2016. <span>“Mendelian Sampling Covariability of Marker Effects and Genetic Values.”</span> <em>Genetics Selection Evolution</em> 48 (1). <a href="https://doi.org/10.1186/s12711-016-0214-0" class="external-link">https://doi.org/10.1186/s12711-016-0214-0</a>.
</div>
<div id="ref-covarrubias-pazaran2016" class="csl-entry">
Covarrubias-Pazaran, Giovanny. 2016. <span>“Genome-Assisted Prediction of Quantitative Traits Using the R Package Sommer.”</span> Edited by Aimin Zhang. <em>PLOS ONE</em> 11 (6): e0156744. <a href="https://doi.org/10.1371/journal.pone.0156744" class="external-link">https://doi.org/10.1371/journal.pone.0156744</a>.
</div>
<div id="ref-Garrick2009" class="csl-entry">
Garrick, Dorian J, Jeremy F Taylor, and Rohan L Fernando. 2009. <span>“Deregressing Estimated Breeding Values and Weighting Information for Genomic Regression Analyses.”</span> <em>Genetics Selection Evolution</em> 41 (1). <a href="https://doi.org/10.1186/1297-9686-41-55" class="external-link">https://doi.org/10.1186/1297-9686-41-55</a>.
</div>
<div id="ref-Lehermeier2017" class="csl-entry">
Lehermeier, Christina, Simon Teyssèdre, and Chris-Carolin Schön. 2017. <span>“Genetic Gain Increases by Applying the Usefulness Criterion with Improved Variance Prediction in Selection of Crosses.”</span> <em>Genetics</em> 207 (4): 1651–61. <a href="https://doi.org/10.1534/genetics.117.300403" class="external-link">https://doi.org/10.1534/genetics.117.300403</a>.
</div>
<div id="ref-Neyhart2019" class="csl-entry">
Neyhart, Jeffrey L, Aaron J Lorenz, and Kevin P Smith. 2019. <span>“Multi-Trait Improvement by Predicting Genetic Correlations in Breeding Crosses.”</span> <em>G3 Genes|Genomes|Genetics</em> 9 (10): 3153–65. <a href="https://doi.org/10.1534/g3.119.400406" class="external-link">https://doi.org/10.1534/g3.119.400406</a>.
</div>
<div id="ref-Toro2010" class="csl-entry">
Toro, Miguel A, and Luis Varona. 2010. <span>“A Note on Mate Allocation for Dominance Handling in Genomic Selection.”</span> <em>Genetics Selection Evolution</em> 42 (1). <a href="https://doi.org/10.1186/1297-9686-42-33" class="external-link">https://doi.org/10.1186/1297-9686-42-33</a>.
</div>
<div id="ref-Varona2018" class="csl-entry">
Varona, Luis, Andres Legarra, Miguel A. Toro, and Zulma G. Vitezica. 2018. <span>“Non-Additive Effects in Genomic Selection.”</span> <em>Frontiers in Genetics</em> 9 (March). <a href="https://doi.org/10.3389/fgene.2018.00078" class="external-link">https://doi.org/10.3389/fgene.2018.00078</a>.
</div>
<div id="ref-Vitezica2013" class="csl-entry">
Vitezica, Zulma G, Luis Varona, and Andres Legarra. 2013. <span>“On the Additive and Dominant Variance and Covariance of Individuals Within the Genomic Selection Scope.”</span> <em>Genetics</em> 195 (4): 1223–30. <a href="https://doi.org/10.1534/genetics.113.155176" class="external-link">https://doi.org/10.1534/genetics.113.155176</a>.
</div>
<div id="ref-werner2020" class="csl-entry">
Werner, Christian R., R. Chris Gaynor, Daniel J. Sargent, Alessandra Lillo, Gregor Gorjanc, and John M. Hickey. 2020. <span>“Genomic Selection Strategies for Clonally Propagated Crops.”</span> <a href="http://dx.doi.org/10.1101/2020.06.15.152017" class="external-link">http://dx.doi.org/10.1101/2020.06.15.152017</a>.
</div>
<div id="ref-Wolfe2021" class="csl-entry">
Wolfe, Marnin D, Ariel W Chan, Peter Kulakow, Ismail Rabbi, and Jean-Luc Jannink. 2021. <span>“Genomic Mating in Outbred Species: Predicting Cross Usefulness with Additive and Total Genetic Covariance Matrices.”</span> <em>Genetics</em> 219 (3). <a href="https://doi.org/10.1093/genetics/iyab122" class="external-link">https://doi.org/10.1093/genetics/iyab122</a>.
</div>
<div id="ref-Xiang2016" class="csl-entry">
Xiang, Tao, Ole Fredslund Christensen, Zulma Gladis Vitezica, and Andres Legarra. 2016. <span>“Genomic Evaluation by Including Dominance Effects and Inbreeding Depression for Purebred and Crossbred Performance with an Application in Pigs.”</span> <em>Genetics Selection Evolution</em> 48 (1). <a href="https://doi.org/10.1186/s12711-016-0271-4" class="external-link">https://doi.org/10.1186/s12711-016-0271-4</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Marnin Wolfe.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
