<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started predicting crosses • genomicMateSelectR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting started predicting crosses">
<meta property="og:description" content="genomicMateSelectR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genomicMateSelectR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/non_additive_models.html">Genomic predictions with non-additive effects</a>
    </li>
    <li>
      <a href="../articles/start_here.html">Getting started predicting crosses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wolfemd/genomicMateSelectR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="start_here_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting started predicting crosses</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/wolfemd/genomicMateSelectR/blob/master/vignettes/start_here.Rmd"><code>vignettes/start_here.Rmd</code></a></small>
      <div class="hidden name"><code>start_here.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wolfemd/genomicMateSelectR">genomicMateSelectR</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: tibble</span>
<span class="co">#&gt; Loading required package: dplyr</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="co">#&gt; Loading required package: tidyr</span>
<span class="co">#&gt; Loading required package: purrr</span></code></pre></div>
<p>The following will be a demonstration of the core functions for predicting mate selection criteria available in <code>genomicMateSelectR</code>.</p>
<div id="simulated-data" class="section level1">
<h1 class="hasAnchor">
<a href="#simulated-data" class="anchor"></a>Simulated data</h1>
<p>I generated simulated data, included in the package (accessible via <code><a href="https://rdrr.io/r/utils/data.html">data()</a></code>), which should clarify the input format(s) required.</p>
<p>The example input data will be explained and used in the vignette below.</p>
<p>Simulation was conducted with the package <code>AlphaSimR</code>.</p>
<ul>
<li>Simulate 100 individuals, 2 chrom., 100 seg. sites, 50 SNP/chr.</li>
<li>2 traits, gen. corr. 0.5, <span class="math inline">\(h^2=0.5\)</span>, 10 QTL/chr. equal parts additive + dominance.</li>
<li>2 cycles of phenotypic selection, based on one of the two traits, top 5 individuals, 10 crosses x 10 progeny</li>
<li>Generates 300 individuals across 3 generations</li>
</ul>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h1>
<p>In a subsequent section, we will dig into the details and more fully explore all the available options in <code>genomicMateSelectR</code>. First, a quick demo of the <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> function using (mostly) ready-to-go pre-prepared input data.</p>
<p>To predict cross variance, 4 inputs are required:</p>
<ol style="list-style-type: decimal">
<li>Phased haplotypes of candidate parents</li>
</ol>
<ul>
<li>Also a matrix of allele dosages</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>Recombination frequency matrix (derived from a genetic map)</li>
<li>Marker effects</li>
<li>Crosses to be predicted</li>
</ol>
<div id="haplotype-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#haplotype-matrix" class="anchor"></a>Haplotype matrix</h2>
<p>First input is a matrix,<code>haploMat</code> with two rows per individual and one column per SNP locus.</p>
<p>Alleles coded <code>{0,1}</code>. <code>rownames(haploMat)</code> contains sample IDs (“GID”) with haplotypes distinguished by the mandatory suffixes “_HapA” and “_HapB”. <code>colnames(haploMat)</code> correspond to SNP IDs (<em>i.e.</em> <span class="math inline">\(chr \_ pos\)</span>), currently requires that chromosomes be distinguished using integers; no prefix to SNP IDs is allowed (i.e. 1_2 <em>not</em> chr1_2 and also not anything like A_2). In this example, the GIDs are all numbers, they are from simulated data; in practice they can be any alphanumeric string.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">haploMat</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">haploMat</span><span class="op">)</span>
<span class="co">#&gt;  int [1:600, 1:100] 0 1 1 1 1 1 1 0 0 1 ...</span>
<span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   ..$ : chr [1:600] "1_HapA" "1_HapB" "2_HapA" "2_HapB" ...</span>
<span class="co">#&gt;   ..$ : chr [1:100] "1_2" "1_3" "1_6" "1_7" ...</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">haploMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span>
<span class="co">#&gt;        1_2 1_3 1_6 1_7 1_8 1_11 1_14 1_17</span>
<span class="co">#&gt; 1_HapA   0   0   1   0   0    0    1    1</span>
<span class="co">#&gt; 1_HapB   1   0   1   0   0    0    0    0</span>
<span class="co">#&gt; 2_HapA   1   0   1   0   0    0    0    0</span>
<span class="co">#&gt; 2_HapB   1   0   0   1   1    0    0    1</span>
<span class="co">#&gt; 3_HapA   1   0   1   0   0    0    0    0</span>
<span class="co">#&gt; 3_HapB   1   0   1   0   0    0    1    1</span></code></pre></div>
<div id="dosages" class="section level3">
<h3 class="hasAnchor">
<a href="#dosages" class="anchor"></a>Dosages</h3>
<p>For now, to use <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>, both a <code>haploMat</code> <em>and</em> another, related matrix (<code>dosages</code>) are required as input.</p>
<p>The matrix <code>dosages</code> has one row per individual sample and one column per SNP marker. Column-names are the same as for <code>haploMat</code>. rownames are the same prefix as in <code>haploMat</code>, but no suffix (“_HapA” or “_HapB”). Genotypes are coded <code>{0, 1, 2}</code> counting the number of (usually) the alternative allele. Important that the counted allele is the same allele as indicated by the value <code>1</code> in the <code>haploMat</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"doseMat"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">doseMat</span><span class="op">)</span>
<span class="co">#&gt;  int [1:300, 1:100] 1 2 2 1 1 0 1 1 1 2 ...</span>
<span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   ..$ : chr [1:300] "1" "2" "3" "4" ...</span>
<span class="co">#&gt;   ..$ : chr [1:100] "1_2" "1_3" "1_6" "1_7" ...</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">doseMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span>
<span class="co">#&gt;   1_2 1_3 1_6 1_7 1_8 1_11 1_14 1_17</span>
<span class="co">#&gt; 1   1   0   2   0   0    0    1    1</span>
<span class="co">#&gt; 2   2   0   1   1   1    0    0    1</span>
<span class="co">#&gt; 3   2   0   2   0   0    0    1    1</span></code></pre></div>
<p><strong>NOTE:</strong> The dual <code>dosages</code> and <code>haploMat</code> input requirement is because most users will be predicting cross means + variances and currently, the function <code><a href="../reference/predCrossMeans.html">predCrossMeans()</a></code>, which is used internally, requires <code>dosages</code> and <code><a href="../reference/predCrossVars.html">predCrossVars()</a></code> needs a <code>haploMat</code>. Its computationally more efficient (I think) to avoid converting haplotype matrices into dosage matrices internally, thus I demand both be specified. On the improvement to-do list.</p>
</div>
</div>
<div id="recombination-frequency-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#recombination-frequency-matrix" class="anchor"></a>Recombination frequency matrix</h2>
<p>The second input, <code>recombFreqMat</code> is a square-symmetric matrix, with dimension <em>p</em>-markers by <em>p</em>-markers. The elements of <code>recombFreqMat</code>, should actually be <code>1-2*recombFreqMat</code>, meaning 1 minus 2 times the expected frequency of recombination between pairs of loci; a decision made for computational efficiency.</p>
<p>The usual starting point to create this matrix will be a centimorgan-scale genetic map. <code>genomicMateSelectR</code> provides a helper function <code><a href="../reference/genmap2recombfreq.html">genmap2recombfreq()</a></code> to facilitate the conversion.</p>
<p>Users can input their genetic map as a named, numeric vector as shown below (<code>genmap</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">genmap</span><span class="op">)</span>; 
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">genmap</span><span class="op">)</span>
<span class="co">#&gt;  Named num [1:100] 0.366 1.014 5.598 6.46 7.216 ...</span>
<span class="co">#&gt;  - attr(*, "names")= chr [1:100] "1_2" "1_3" "1_6" "1_7" ...</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genmap</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="co">#&gt;        1_2        1_3        1_6        1_7        1_8       1_11       1_14 </span>
<span class="co">#&gt;  0.3664138  1.0141046  5.5980510  6.4599211  7.2155092  9.0206727 12.7646012 </span>
<span class="co">#&gt;       1_17       1_20       1_21 </span>
<span class="co">#&gt; 14.2529985 16.2965964 16.8133229</span></code></pre></div>
<p>The input format for <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>, for reasons of computational efficiency, is actually <code>1-2*recombFreqMat</code> where <code>recombFreqMat</code> is a</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/genmap2recombfreq.html">genmap2recombfreq</a></span><span class="op">(</span><span class="va">genmap</span>, nChr <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt;  num [1:100, 1:100] 0 0.363 0.5 0.5 0.5 ...</span>
<span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   ..$ : chr [1:100] "1_2" "1_3" "1_6" "1_7" ...</span>
<span class="co">#&gt;   ..$ : chr [1:100] "1_2" "1_3" "1_6" "1_7" ...</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">recombFreqMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;           1_2       1_3       1_6       1_7       1_8</span>
<span class="co">#&gt; 1_2 0.0000000 0.3631033 0.4999857 0.4999975 0.4999994</span>
<span class="co">#&gt; 1_3 0.3631033 0.0000000 0.4999478 0.4999907 0.4999979</span>
<span class="co">#&gt; 1_6 0.4999857 0.4999478 0.0000000 0.4108012 0.4803183</span>
<span class="co">#&gt; 1_7 0.4999975 0.4999907 0.4108012 0.0000000 0.3896748</span>
<span class="co">#&gt; 1_8 0.4999994 0.4999979 0.4803183 0.3896748 0.0000000</span></code></pre></div>
<p>Now calc. <code>1-2*recombFreqMat</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fl">1</span><span class="op">-</span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">recombFreqMat</span><span class="op">)</span>
<span class="va">recombFreqMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;              1_2          1_3          1_6          1_7          1_8</span>
<span class="co">#&gt; 1_2 1.000000e+00 2.737934e-01 2.856654e-05 5.096204e-06 1.124479e-06</span>
<span class="co">#&gt; 1_3 2.737934e-01 1.000000e+00 1.043361e-04 1.861332e-05 4.107035e-06</span>
<span class="co">#&gt; 1_6 2.856654e-05 1.043361e-04 1.000000e+00 1.783976e-01 3.936349e-02</span>
<span class="co">#&gt; 1_7 5.096204e-06 1.861332e-05 1.783976e-01 1.000000e+00 2.206503e-01</span>
<span class="co">#&gt; 1_8 1.124479e-06 4.107035e-06 3.936349e-02 2.206503e-01 1.000000e+00</span></code></pre></div>
<p>This corresponds to the input to <code>predictCrosses(recombFreqMat=)</code> and the example dataset included with the package: <code>data("recombFreqMat")</code></p>
</div>
<div id="marker-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#marker-effects" class="anchor"></a>Marker effects</h2>
<p>The third input, for <code>predictCrosses(snpeffs=)</code> are marker-effects from genome-wide marker regression e.g. RR-BLUP.</p>
<p>Marker-effects are supplied as a <code>tibble</code> with one row per trait, one chr-type column named <strong>Trait</strong> and one or more list-type columns containing SNP-effects. Each element of each list-type SNP-effects column corresponds to effects for a single trait, and contains a matrix with a single column, rownames corresponding to SNP IDs and should match the IDs in <code>recombFreqMat</code> and <code>haploMat</code>.</p>
<p>The number and required column names for the SNP-effect list-type columns depends on the value of <code>modelType==</code>. There are 3 model types implemented <code>"A"</code>, <code>"AD"</code> and <code>"DirDom"</code>. The <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> function is designed to be run with output produced by running the <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> function using the same <code>modelType</code> setting. Both are designed to ensure / encourage the correct formulation of additive+dominance models.</p>
<p>For this quick example, the additive-only model: <code>modelType="A"</code>. I’ll introduce each model later and explain the correct inputs / expected outputs for those.</p>
<p>The example dataset <code>data(snpeffsA)</code> is <code>gpredsA$genomicPredOut[[1]]</code> where <code>gpreds</code> is the output of running <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> with <code>modelType = "A"</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">snpeffsA</span><span class="op">)</span>
<span class="va">snpeffsA</span>
<span class="co">#&gt; # A tibble: 2 × 5</span>
<span class="co">#&gt;   Trait  gblups             varcomps     fixeffs      allelesubsnpeff</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;list&gt;             &lt;list&gt;       &lt;list&gt;       &lt;list&gt;         </span>
<span class="co">#&gt; 1 Trait1 &lt;tibble [300 × 2]&gt; &lt;df [2 × 4]&gt; &lt;df [1 × 5]&gt; &lt;dbl [100 × 1]&gt;</span>
<span class="co">#&gt; 2 Trait2 &lt;tibble [300 × 2]&gt; &lt;df [2 × 4]&gt; &lt;df [1 × 5]&gt; &lt;dbl [100 × 1]&gt;</span></code></pre></div>
<p>Because <code>snpeffsA</code> is the output of <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code>, there are columns that are not required and will be ignored by <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>. The unecessary columns are <strong>gblups</strong>, <strong>varcomps</strong> and <strong>fixeffs</strong>.</p>
<p>Required:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snpeffsA</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Trait</span>,<span class="va">allelesubsnpeff</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 2 × 2</span>
<span class="co">#&gt;   Trait  allelesubsnpeff</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;list&gt;         </span>
<span class="co">#&gt; 1 Trait1 &lt;dbl [100 × 1]&gt;</span>
<span class="co">#&gt; 2 Trait2 &lt;dbl [100 × 1]&gt;</span></code></pre></div>
<p>Two rows, one for each trait. One list-column, required to be labelled <strong>allelsubsnpeff</strong>, containing a single-column matrix of SNP-effects, rownames identifying SNP IDs.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snpeffsA</span><span class="op">$</span><span class="va">allelesubsnpeff</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt;            [,1]</span>
<span class="co">#&gt; 1_2  14.6901150</span>
<span class="co">#&gt; 1_3   0.4877389</span>
<span class="co">#&gt; 1_6   4.3651094</span>
<span class="co">#&gt; 1_7  -1.9612765</span>
<span class="co">#&gt; 1_8  -2.2205335</span>
<span class="co">#&gt; 1_11  2.7950027</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">snpeffsA</span><span class="op">$</span><span class="va">allelesubsnpeff</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;        V1          </span>
<span class="co">#&gt;  Min.   :-17.6166  </span>
<span class="co">#&gt;  1st Qu.: -4.1909  </span>
<span class="co">#&gt;  Median :  0.4571  </span>
<span class="co">#&gt;  Mean   :  0.6947  </span>
<span class="co">#&gt;  3rd Qu.:  5.3649  </span>
<span class="co">#&gt;  Max.   : 27.7530</span></code></pre></div>
</div>
<div id="crosses-to-predict" class="section level2">
<h2 class="hasAnchor">
<a href="#crosses-to-predict" class="anchor"></a>Crosses to predict</h2>
<p>Lastly, we just need to tell <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>, which crosses to predict. The input to argument <code>CrossesToPredict=</code> should be a <code>tibble</code> or <code>data.frame</code> with two-columns. Columns should be character-class (no factors!), column names should be <strong>sireID</strong> and <strong>damID</strong>. Names should be present in the rownames of <code>haploMat</code> and the row/colnames of <code>recombFreqMat</code>.</p>
<p>There is a helper function <code><a href="../reference/crosses2predict.html">crosses2predict()</a></code> that takes a vector of parents as input and creates a <code>data.frame</code> of all pairwise matings, including self-crosses, but <em>excluding</em> reciprocal crosses, i.e. use as male == use as female. Basically <code><a href="../reference/crosses2predict.html">crosses2predict()</a></code> specifies crosses on the diagonal and in the upper-triangle of a potentially square-symmetrical mating matrix.</p>
<p>Here, I’ll just randomly choose 5 parents and make a list of the 15 <code>CrossesToPredict</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span>;
<span class="va">parents</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">snpeffsA</span><span class="op">$</span><span class="va">gblups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">GID</span>, size <span class="op">=</span> <span class="fl">5</span>, replace <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">CrossesToPredict</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/crosses2predict.html">crosses2predict</a></span><span class="op">(</span><span class="va">parents</span><span class="op">)</span>
<span class="va">CrossesToPredict</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt; # A tibble: 6 × 2</span>
<span class="co">#&gt;   sireID damID</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1 49     49   </span>
<span class="co">#&gt; 2 49     153  </span>
<span class="co">#&gt; 3 49     74   </span>
<span class="co">#&gt; 4 49     228  </span>
<span class="co">#&gt; 5 49     146  </span>
<span class="co">#&gt; 6 153    153</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">CrossesToPredict</span><span class="op">)</span>
<span class="co">#&gt; tibble [15 × 2] (S3: tbl_df/tbl/data.frame)</span>
<span class="co">#&gt;  $ sireID: chr [1:15] "49" "49" "49" "49" ...</span>
<span class="co">#&gt;  $ damID : chr [1:15] "49" "153" "74" "228" ...</span></code></pre></div>
</div>
<div id="predict-crosses" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-crosses" class="anchor"></a>Predict crosses</h2>
<p>And behold, we have the stuff we need to predict crosses.</p>
<p><strong>Selection Indices:</strong><br><code>genomicMateSelectR</code> is built to facilitate mating choices based on multiple traits, particularly using a user-specified set of weights (<strong>SIwts</strong>) to produce cross merit predictions on a linear selection index (<code>selInd</code>). When supplied multiple traits, <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> will predict both the variances of those traits and their covariance in each family. When the user specifies <code>selInd = T</code> and supplies a vector of weights with names matching the trait names, <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> will add predictions for a new composite trait, <strong>SELIND</strong> to the output.</p>
<p><strong>SIwts</strong> are used as-is by <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> so the user is responsible for choosing “good” values; a topic beyond the scope of this manual / vignette.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SIwts</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.75</span>,<span class="fl">0.25</span><span class="op">)</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu">`names&lt;-`</span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Trait1"</span>,<span class="st">"Trait2"</span><span class="op">)</span><span class="op">)</span>
<span class="va">SIwts</span>
<span class="co">#&gt; Trait1 Trait2 </span>
<span class="co">#&gt;   0.75   0.25</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crossPreds</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/predictCrosses.html">predictCrosses</a></span><span class="op">(</span>modelType<span class="op">=</span><span class="st">"A"</span>,
                           selInd <span class="op">=</span> <span class="cn">T</span>, SIwts <span class="op">=</span> <span class="va">SIwts</span>,
                           CrossesToPredict<span class="op">=</span><span class="va">CrossesToPredict</span>,
                           snpeffs<span class="op">=</span><span class="va">snpeffsA</span>, 
                           dosages<span class="op">=</span><span class="va">doseMat</span>,
                           haploMat<span class="op">=</span><span class="va">haploMat</span>,
                           recombFreqMat<span class="op">=</span><span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt; [1] "Predicting cross variance parameters"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.07 mins for 15 crosses"</span>
<span class="co">#&gt; [1] "Predicting cross means"</span>
<span class="co">#&gt; [1] "Computing SELECTION INDEX means and variances."</span></code></pre></div>
<p>Now we can quickly examine the output.</p>
<p>Produces a single-row tibble (basically a 2 element list) with two columns <strong>tidyPreds</strong> and <strong>rawPreds</strong>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crossPreds</span>
<span class="co">#&gt; # A tibble: 1 × 2</span>
<span class="co">#&gt;   tidyPreds         rawPreds        </span>
<span class="co">#&gt;   &lt;list&gt;            &lt;list&gt;          </span>
<span class="co">#&gt; 1 &lt;tibble [45 × 9]&gt; &lt;named list [2]&gt;</span></code></pre></div>
<p>Most users and breeders will work with the <strong>tidyPreds</strong>. This simplified output was designed especially for those interested in selections based on <strong>SELIND</strong>.</p>
<p>Each row contains relevant predictions for a single trait in a single cross. Trait <strong><em>co</em></strong>variances are excluded from this output in order to show both mean (<strong>predMean</strong>), variance / standard deviation (<strong>predVar</strong>, <strong>predSD</strong>) and usefulness (<strong>predUsefulness</strong>) side-by-side.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crossPreds</span><span class="op">$</span><span class="va">tidyPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 45 × 9</span>
<span class="co">#&gt;    sireID damID Nsegsnps predOf Trait  predMean predVar predSD predUsefulness</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1 49     49          22 BV     SELIND    115.     361.   19.0         166.  </span>
<span class="co">#&gt;  2 49     49          22 BV     Trait1    194.     707.   26.6         265.  </span>
<span class="co">#&gt;  3 49     49          22 BV     Trait2   -119.    1519.   39.0         -15.1 </span>
<span class="co">#&gt;  4 49     153         46 BV     SELIND    108.     454.   21.3         165.  </span>
<span class="co">#&gt;  5 49     153         46 BV     Trait1    174.    1079.   32.8         262.  </span>
<span class="co">#&gt;  6 49     153         46 BV     Trait2    -89.4   1813.   42.6          24.3 </span>
<span class="co">#&gt;  7 49     74          48 BV     SELIND     97.5    458.   21.4         155.  </span>
<span class="co">#&gt;  8 49     74          48 BV     Trait1    165.     974.   31.2         249.  </span>
<span class="co">#&gt;  9 49     74          48 BV     Trait2   -106.    1510.   38.9          -2.05</span>
<span class="co">#&gt; 10 49     228         53 BV     SELIND     97.3    536.   23.2         159.  </span>
<span class="co">#&gt; # … with 35 more rows</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crossPreds</span><span class="op">$</span><span class="va">tidyPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> <span class="va">str</span>
<span class="co">#&gt; tibble [45 × 9] (S3: tbl_df/tbl/data.frame)</span>
<span class="co">#&gt;  $ sireID        : chr [1:45] "49" "49" "49" "49" ...</span>
<span class="co">#&gt;  $ damID         : chr [1:45] "49" "49" "49" "153" ...</span>
<span class="co">#&gt;  $ Nsegsnps      : int [1:45] 22 22 22 46 46 46 48 48 48 53 ...</span>
<span class="co">#&gt;  $ predOf        : chr [1:45] "BV" "BV" "BV" "BV" ...</span>
<span class="co">#&gt;  $ Trait         : chr [1:45] "SELIND" "Trait1" "Trait2" "SELIND" ...</span>
<span class="co">#&gt;  $ predMean      : num [1:45] 115 194 -119 108 174 ...</span>
<span class="co">#&gt;  $ predVar       : num [1:45] 361 707 1519 454 1079 ...</span>
<span class="co">#&gt;  $ predSD        : num [1:45] 19 26.6 39 21.3 32.8 ...</span>
<span class="co">#&gt;  $ predUsefulness: num [1:45] 166.2 264.7 -15.1 164.9 261.6 ...</span></code></pre></div>
<p>Additional columns in <strong>tidyPreds</strong>:</p>
<ul>
<li><p><code>Nsegsnps</code>: the number of SNPs expected to segregate in this cross. For compute efficiency, when predicting cross variances, <code>genomicMateSelectR</code> detects which SNPs will actually cause genetic variance within the cross, and restricts the analysis to those loci.</p></li>
<li><p><code>predOf</code>: “BV” for breeding value (all models), “TGV” for total genetic values (models “AD” and “DirDom” only). Here only “BV” because <code>modelType="A"</code>.</p></li>
<li><p><code>predUsefulness</code> is <code>predMean + stdSelInt*predSD</code>, where <code>stdSelInt</code> is the standardized selection intensity input to <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code>, <code>stdSelInt = 2.67</code> by default.</p></li>
</ul>
<p>Complete details of the predictions, especially including trait-trait covariance predictions are available in the two-element, named-list contained in <strong>rawPreds</strong>:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crossPreds</span><span class="op">$</span><span class="va">rawPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> <span class="va">str</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ predMeans:List of 1</span>
<span class="co">#&gt;   ..$ : tibble [30 × 7] (S3: tbl_df/tbl/data.frame)</span>
<span class="co">#&gt;   .. ..$ Trait   : chr [1:30] "Trait1" "Trait1" "Trait1" "Trait1" ...</span>
<span class="co">#&gt;   .. ..$ sireID  : chr [1:30] "49" "49" "49" "49" ...</span>
<span class="co">#&gt;   .. ..$ damID   : chr [1:30] "49" "153" "74" "228" ...</span>
<span class="co">#&gt;   .. ..$ sireGEBV: num [1:30] 194 194 194 194 194 ...</span>
<span class="co">#&gt;   .. ..$ damGEBV : num [1:30] 194 154 137 111 111 ...</span>
<span class="co">#&gt;   .. ..$ predOf  : chr [1:30] "MeanBV" "MeanBV" "MeanBV" "MeanBV" ...</span>
<span class="co">#&gt;   .. ..$ predMean: num [1:30] 194 174 165 152 152 ...</span>
<span class="co">#&gt;  $ predVars :List of 1</span>
<span class="co">#&gt;   ..$ : tibble [45 × 8] (S3: tbl_df/tbl/data.frame)</span>
<span class="co">#&gt;   .. ..$ sireID     : chr [1:45] "49" "49" "49" "49" ...</span>
<span class="co">#&gt;   .. ..$ damID      : chr [1:45] "49" "49" "49" "153" ...</span>
<span class="co">#&gt;   .. ..$ Nsegsnps   : int [1:45] 22 22 22 46 46 46 48 48 48 53 ...</span>
<span class="co">#&gt;   .. ..$ ComputeTime: Named num [1:45] 0.268 0.268 0.268 0.246 0.246 ...</span>
<span class="co">#&gt;   .. .. ..- attr(*, "names")= chr [1:45] "elapsed" "elapsed" "elapsed" "elapsed" ...</span>
<span class="co">#&gt;   .. ..$ Trait1     : chr [1:45] "Trait1" "Trait2" "Trait1" "Trait1" ...</span>
<span class="co">#&gt;   .. ..$ Trait2     : chr [1:45] "Trait1" "Trait2" "Trait2" "Trait1" ...</span>
<span class="co">#&gt;   .. ..$ predOf     : chr [1:45] "VarBV" "VarBV" "VarBV" "VarBV" ...</span>
<span class="co">#&gt;   .. ..$ predVar    : num [1:45] 707 1519 -350 1079 1813 ...</span></code></pre></div>
<p>At <code>crossPreds$rawPreds[[1]]$predMeans</code> are the predictions for cross means, and at <code>crossPreds$rawPreds[[1]]$predVars</code> are the predictions for cross variances.</p>
<p>For sake of example, we should make a plot (predMean vs. predSD), for “fun” and then make a selection of the top 5 crosses, based on their predicted usefulness on the selection index:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crossPreds</span><span class="op">$</span><span class="va">tidyPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Trait</span><span class="op">==</span><span class="st">"SELIND"</span><span class="op">)</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">predMean</span>,y<span class="op">=</span><span class="va">predSD</span>,
                                 color<span class="op">=</span><span class="va">predUsefulness</span>, size<span class="op">=</span><span class="va">predUsefulness</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="start_here_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crossPreds</span><span class="op">$</span><span class="va">tidyPreds</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Trait</span><span class="op">==</span><span class="st">"SELIND"</span><span class="op">)</span> <span class="op"><a href="../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">predUsefulness</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 5 × 9</span>
<span class="co">#&gt;   sireID damID Nsegsnps predOf Trait  predMean predVar predSD predUsefulness</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1 49     49          22 BV     SELIND    115.     361.   19.0           166.</span>
<span class="co">#&gt; 2 49     153         46 BV     SELIND    108.     454.   21.3           165.</span>
<span class="co">#&gt; 3 153    153         33 BV     SELIND    101.     546.   23.4           163.</span>
<span class="co">#&gt; 4 49     228         53 BV     SELIND     97.3    536.   23.2           159.</span>
<span class="co">#&gt; 5 153    228         55 BV     SELIND     89.9    628.   25.1           157.</span>
  <span class="co"># arrange(desc(predUsefulness)) %&gt;% slice(1:5) # same as `slice_max`</span></code></pre></div>
</div>
</div>
<div id="next-steps" class="section level1">
<h1 class="hasAnchor">
<a href="#next-steps" class="anchor"></a>Next steps</h1>
<p>That concludes the quick demo, focused on input formatting and output contents, using the basic additive-model in the function <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> to get selection-index cross usefulness predictions.</p>
<div id="additional-vignettes" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-vignettes" class="anchor"></a>Additional vignettes</h2>
<p>In subsequent vignettes I will provide:</p>
<ol style="list-style-type: decimal">
<li>Learn the <a href="articles/non_additive_models.html">features for incorporating non-additive effects into predictions</a>: <code>modelType="AD"</code>, <code>modelType="DirDom</code> and their correct useage. Includes a theoretical introduction, followed by a demonstration of the use of <code><a href="../reference/runGenomicPredictions.html">runGenomicPredictions()</a></code> and <code><a href="../reference/predictCrosses.html">predictCrosses()</a></code> to accomplish predictions of individual and cross performances.</li>
</ol>
<ul>
<li>Use of the functions for cross-validation (<code><a href="../reference/runCrossVal.html">runCrossVal()</a></code>, <code><a href="../reference/runParentWiseCrossVal.html">runParentWiseCrossVal()</a></code>,</li>
<li>possibly the imputation and breeding data processing helper functions (wrappers for e.g. Beagle and code for BreedBase data cleaning).</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Marnin Wolfe.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
