<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run genomic predictions — runGenomicPredictions • genomicMateSelectR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run genomic predictions — runGenomicPredictions"><meta property="og:description" content='Run GBLUP model using mmer, potentially on multiple
traits. Returns genomic BLUPs (GEBV and GETGV). If requested, returns
backsolved marker effects (equivalent to ridge regression / SNP-BLUP).
Three models are
enabled: additive-only ("A"), additive-plus-dominance ("AD") and a
directional-dominance model that incorporates a genome-wide homozygosity
effect ("DirDom"). Inbreeding effect is included in output GEBV/GETGV
predictions *after* backsolving SNP effects. If requested, returns
GEBV/GETGV computed for a selection index using selInd=TRUE
and supplying SIwts.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">genomicMateSelectR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/non_additive_models.html">Genomic predictions with non-additive effects</a>
    </li>
    <li>
      <a href="../articles/start_here.html">Getting started predicting crosses</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/wolfemd/genomicMateSelectR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run genomic predictions</h1>
    <small class="dont-index">Source: <a href="https://github.com/wolfemd/genomicMateSelectR/blob/HEAD/R/gmsFunctions.R" class="external-link"><code>R/gmsFunctions.R</code></a></small>
    <div class="hidden name"><code>runGenomicPredictions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run GBLUP model using <code><a href="https://rdrr.io/pkg/sommer/man/mmer.html" class="external-link">mmer</a></code>, potentially on multiple
traits. Returns genomic BLUPs (GEBV and GETGV). If requested, returns
backsolved marker effects (equivalent to ridge regression / SNP-BLUP).
Three models are
enabled: additive-only ("A"), additive-plus-dominance ("AD") and a
directional-dominance model that incorporates a genome-wide homozygosity
effect ("DirDom"). Inbreeding effect is included in output GEBV/GETGV
predictions *after* backsolving SNP effects. If requested, returns
GEBV/GETGV computed for a selection index using <code>selInd=TRUE</code>
and supplying <code>SIwts</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">runGenomicPredictions</span><span class="op">(</span>
  <span class="va">modelType</span>,
  <span class="va">selInd</span>,
  SIwts <span class="op">=</span> <span class="cn">NULL</span>,
  getMarkEffs <span class="op">=</span> <span class="cn">FALSE</span>,
  returnPEV <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">blups</span>,
  <span class="va">grms</span>,
  dosages <span class="op">=</span> <span class="cn">NULL</span>,
  gid <span class="op">=</span> <span class="st">"GID"</span>,
  ncores <span class="op">=</span> <span class="fl">1</span>,
  nBLASthreads <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>modelType</dt>
<dd><p>string, "A", "AD", "DirDom".
modelType="A": additive-only, GEBVS
modelType="AD": the "classic" add-dom model, GEBVS+GEDDs = GETGVs
modelType="DirDom":
the "genotypic" add-dom model with prop. homozygous
fit as a fixed-effect, to estimate a genome-wide inbreeding effect.
obtains add-dom effects, computes
allele sub effects (\(\alpha = a + d(q-p)\))
incorporates into GEBV and GETGV</p></dd>
<dt>selInd</dt>
<dd><p>logical, TRUE/FALSE, selection index accuracy estimates,
requires input weights via <code>SIwts</code></p></dd>
<dt>SIwts</dt>
<dd><p>required if <code>selInd=FALSE</code>, named vector of selection
index weights, names match the "Trait" variable in <code>blups</code></p></dd>
<dt>getMarkEffs</dt>
<dd><p>T/F return marker effects, backsolved from GBLUP</p></dd>
<dt>returnPEV</dt>
<dd><p>T/F return PEVs in GBLUP</p></dd>
<dt>blups</dt>
<dd><p>nested data.frame with list-column "TrainingData" containing
BLUPs. Each element of "TrainingData" list, is data.frame with de-regressed
BLUPs, BLUPs and weights (WT) for training and test.</p></dd>
<dt>grms</dt>
<dd><p>list of genomic relation matrices (GRMs, aka kinship matrices).
Any genotypes in the GRMs get predicted with, or without phenotypes.
Each element is named either A or D. Matrices supplied must match
required by A, AD and DirDom models. e.g. grms=list(A=A,D=D).</p></dd>
<dt>dosages</dt>
<dd><p>dosage matrix. required only for modelType=="DirDom".
Also required if getMarkEffs==TRUE.
Assumes SNPs coded 0, 1, 2. Nind rows x Nsnp
cols, numeric matrix, with rownames and colnames to indicate SNP/ind ID</p></dd>
<dt>gid</dt>
<dd><p>string variable name used for genotype ID's in e.g. <code>blups</code> (default="GID")</p></dd>
<dt>ncores</dt>
<dd><p>number of cores</p></dd>
<dt>nBLASthreads</dt>
<dd><p>number of cores for each worker to use for multi-thread BLAS</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>tibble, one row, two list columns (basically a named two-element
list of lists): <code>gblups[[1]]</code> and <code>genomicPredOut[[1]]</code>.
codegblups[[1]]: tibble of predicted GEBV/GETGV, all traits and potentially
SELIND genomic BLUPs along the columns.
<code>genomicPredOut[[1]]</code> is a tibble that contains
some combination of lists-columns:</p><ul><li><p>gblups</p></li>
<li><p>varcomps,</p></li>
<li><p>fixeffs,</p></li>
<li><p>allelesubsnpeff,</p></li>
<li><p>addsnpeff,</p></li>
<li><p>domstar_snpeff,</p></li>
<li><p>domsnpeff</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other prediction_functions: 
<code><a href="predictCrosses.html">predictCrosses</a>()</code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marnin Wolfe.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

